fix galaxy login to work with python3

umeboshi2 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->
I wrote this to workaround issue #43730.  I tried to do this in a way that wouldn't interfere with python2 users, but I'm not familiar enough with ```six``` to do this cleaner, but I instead chose to use 
```sys.version_info``` to adjust the json response from ```bytes``` to ```str```.
 

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
galaxy

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
I pulled from github to make pull request.  Version is from git master.
```
ansible 2.7.0.dev0 (galaxy-login-py3 ae43eca45a) last updated 2018/08/06 14:54:52 (GMT -500)
  config file = None
  configured module search path = ['/freespace/home/umeboshi/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /freespace/home/umeboshi/workspace/A/ansible/lib/ansible
  executable location = /freespace/home/umeboshi/.virtualenvs/testansible/bin/ansible
  python version = 3.5.3 (default, Jan 19 2017, 14:11:04) [GCC 6.3.0 20170118]

```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
We need your Github login to identify you.
This information will not be sent to Galaxy, only to api.github.com.
The password will not be displayed.

Use --github-token if you do not want to enter your password.

Github Username: umeboshi2
Password for umeboshi2: 
Opened /freespace/home/umeboshi/.ansible_galaxy
Successfully logged into Galaxy as umeboshi2
```

umeboshi2 committed:
fix galaxy login to work with python3

umeboshi2 committed:
fix galaxy login to work with python3

umeboshi2 committed:
fix galaxy login to work with python3

umeboshi2 committed:
fix galaxy login to work with python3

ericwb commented on lib/ansible/galaxy/login.py:
```diff
@@ -80,8 +81,14 @@ def remove_github_token(self):
+                              url_password=self.github_password,
+                              force_basic_auth=True,)
+            resp = opened.read()
+            if sys.version_info.major > 2:
```
Six is pretty easy to use here:
 `   if not six.PY2:`

Note: Don't use `if six.PY3` because there might one day be a `six.PY4`

umeboshi2 replied:
Cool!  I've never really used it.  I used the greater than ">" when checking major version in case there would ever be a python4.

umeboshi2 replied:
I went to look at the unit test.  It looks like six.PY3 is being imported but not actually used in the test.  It seems that the only test that is run on "login" is that the option parser correctly parsed the command.  I would like to see where a unit test already exists to to perform a successful login in python2.  I just looked in the ```test/units/cli``` directory, since that seemed to be the most likely place it would be.

Akasurde replied:
```
tokens = json.loads(to_text(opened.read(), errors='surrogate_or_strict'))
```
no need for checking Python version. `to_text` will take care of that.

umeboshi2 replied:
where is ```to_text```?

Akasurde replied:
https://github.com/ansible/ansible/blob/88a738b0bacd32d0027f71682d7965aad3c9a040/lib/ansible/module_utils/_text.py#L169

umeboshi2 replied:
I tried ``to_text``

diff:
```diff --git a/lib/ansible/galaxy/login.py b/lib/ansible/galaxy/login.py
index 6fa221a2e8..775ba8c559 100644
--- a/lib/ansible/galaxy/login.py
+++ b/lib/ansible/galaxy/login.py
@@ -31,6 +31,7 @@ from ansible.module_utils.six.moves import input
 from ansible.module_utils.six.moves.urllib.parse import quote as urlquote, urlparse
 from ansible.module_utils.six.moves.urllib.error import HTTPError
 from ansible.module_utils.urls import open_url
+from ansible.module_utils._text import to_text
 from ansible.utils.color import stringc
 
 try:
@@ -86,9 +87,8 @@ class GalaxyLogin(object):
                               url_password=self.github_password,
                               force_basic_auth=True,)
             resp = opened.read()
-            if sys.version_info.major > 2:
-                resp = resp.decode()
-            tokens = json.loads(resp)
+            tokens = json.loads(to_text(opened.read(),
+                                        errors='surrogate_or_strict'))
         except HTTPError as e:
             res = json.load(e)
             raise AnsibleError(res['message'])
@@ -114,10 +114,8 @@ class GalaxyLogin(object):
                               url_username=self.github_username,
                               url_password=self.github_password,
                               force_basic_auth=True, data=args)
-            resp = opened.read()
-            if sys.version_info.major > 2:
-                resp = resp.decode()
-            data = json.loads(resp)
+            data = json.loads(to_text(opened.read(),
+                                      errors="surrogate_or_strict"))
         except HTTPError as e:
             res = json.load(e)
             raise AnsibleError(res['message'])
```

traceback:
```(testansible) umeboshi@bard:~/workspace/A/ansible$ ansible-galaxy login -vvv
ansible-galaxy 2.7.0.dev0 (galaxy-login-py3 ae43eca45a) last updated 2018/08/06 14:54:52 (GMT -500)
  config file = None
  configured module search path = ['/freespace/home/umeboshi/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']                                                                                       
  ansible python module location = /freespace/home/umeboshi/workspace/A/ansible/lib/ansible
  executable location = /freespace/home/umeboshi/.virtualenvs/testansible/bin/ansible-galaxy
  python version = 3.5.3 (default, Jan 19 2017, 14:11:04) [GCC 6.3.0 20170118]
No config file found; using defaults
Opened /freespace/home/umeboshi/.ansible_galaxy


We need your Github login to identify you.
This information will not be sent to Galaxy, only to api.github.com.
The password will not be displayed.

Use --github-token if you do not want to enter your password.

Github Username: umeboshi2
Password for umeboshi2: 
ERROR! Unexpected Exception, this is probably a bug: Expecting value: line 1 column 1 (char 0)
the full traceback was:

Traceback (most recent call last):
  File "/freespace/home/umeboshi/workspace/A/ansible/bin/ansible-galaxy", line 118, in <module>
    exit_code = cli.run()
  File "/freespace/home/umeboshi/workspace/A/ansible/lib/ansible/cli/galaxy.py", line 162, in run
    self.execute()
  File "/freespace/home/umeboshi/workspace/A/ansible/lib/ansible/cli/__init__.py", line 155, in execute
    fn()
  File "/freespace/home/umeboshi/workspace/A/ansible/lib/ansible/cli/galaxy.py", line 570, in execute_login
    github_token = login.create_github_token()
  File "/freespace/home/umeboshi/workspace/A/ansible/lib/ansible/galaxy/login.py", line 110, in create_github_token
    self.remove_github_token()
  File "/freespace/home/umeboshi/workspace/A/ansible/lib/ansible/galaxy/login.py", line 91, in remove_github_token
    errors='surrogate_or_strict'))
  File "/usr/lib/python3.5/json/__init__.py", line 319, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.5/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Thanks for the suggestions.

ericwb commented on lib/ansible/galaxy/login.py:
```diff
@@ -24,6 +24,7 @@
+import sys
 
 from ansible.errors import AnsibleError, AnsibleOptionsError
 from ansible.module_utils.six.moves import input
```
Add:
`from ansible.module_utils import six`

ericwb commented on lib/ansible/galaxy/login.py:
```diff
@@ -103,8 +110,14 @@ def create_github_token(self):
+                              url_password=self.github_password,
+                              force_basic_auth=True, data=args)
+            resp = opened.read()
+            if sys.version_info.major > 2:
```
same here, use `if not six.PY2`

Akasurde replied:
same here.

ericwb reviewed (changes requested):
Minor changes to use six module.   Plus, you should also add a unit test.

umeboshi2 committed:
use six instead of sys and handle errors

umeboshi2 committed:
use six instead of sys and handle errors

umeboshi2 committed:
use six instead of sys and handle errors

umeboshi2 committed:
use six instead of sys and handle errors

umeboshi2 committed:
dry up handling http error ; add fixme comments to help fix bare except statements

umeboshi2 committed:
dry up handling http error ; add fixme comments to help fix bare except statements

umeboshi2 committed:
dry up handling http error ; add fixme comments to help fix bare except statements

umeboshi2 committed:
oops, fix undefined variable

umeboshi2 committed:
oops, fix undefined variable

umeboshi2 committed:
oops, fix undefined variable

umeboshi2 committed:
test ci

umeboshi2 committed:
test ci

umeboshi2 committed:
test ci

umeboshi2 committed:
Merge branch 'devel' into galaxy-login-py3

dagwieers committed:
HP-UX: Fix ansible_pkg_mgr to use swdepot (#51411)

* HP-UX: Fix ansible_pkg_mgr to use swdepot

* Revert to use /usr/sbin/swlist

umeboshi2 committed:
Merge branch 'devel' into galaxy-login-py3

dagwieers committed:
Update Docsite edit (#51115)

* Improved template for docsite edits

rluisr committed:
add example into os_server for server group (#51375)

* add example for server group

kysamson committed:
Clarify priority setting in Inventory doc (#51343)

* priority setting note in inventory doc

The `ansible_group_priority` variable is also described in *Using Variables* and was given a clarification in issue #41512. The variable is also documented in *Working with Inventory* without the caveat that it needs to be declared on the inventory and not a group_vars file. This change adds that as a note

+label: docsite_pr

maxamillion committed:
dnf load substitutions from installroot (#51206)

Fixes #51059

Signed-off-by: Adam Miller <admiller@redhat.com>

umeboshi2 committed:
Merge branch 'galaxy-login-py3' of github.com:umeboshi2/ansible into galaxy-login-py3

umeboshi2 committed:
Merge branch 'galaxy-login-py3' of github.com:umeboshi2/ansible into galaxy-login-py3

umeboshi2 commented:
http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul
